File: webf/lib/src/module/event_source.dart
/*
 * Copyright (C) 2024-present The OpenWebF Company. All rights reserved.
 * Licensed under GNU GPL with Enterprise exception.
 */
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:dio/dio.dart';
import 'package:dio/io.dart';
import 'package:webf/dom.dart';
import 'package:webf/foundation.dart';
import 'package:webf/launcher.dart';
import 'package:webf/module.dart';

typedef EventSourceEventCallback = void Function(String id, Event event);

class _EventSourceConnection {
  StreamSubscription<String>? subscription;
  CancelToken? cancelToken;
  HttpClientRequest? httpRequest;
  Timer? reconnectTimer;
  String? lastEventId;
  int retryMs = 3000;
  bool closed = false;
  String currentEventType = 'message';
  StringBuffer dataBuffer = StringBuffer();
}

class EventSourceModule extends WebFBaseModule {
  @override
  String get name => 'EventSource';

  int _clientId = 0;
  final Map<String, _EventSourceConnection> _connectionMap = {};
  final Map<String, Map<String, bool>> _listenMap = {};

  EventSourceModule(super.moduleManager);

  @override
  String invoke(String method, List<dynamic> params) {
    switch (method) {
      case 'init':
        final String url = params[0];
        final bool withCredentials = params[1] == true || params[1] == 'true';
        return _init(url, withCredentials);
      case 'close':
        _close(params[0]);
        return '';
      case 'addEvent':
        _addEvent(params[0], params[1]);
        return '';
    }
    return '';
  }

  @override
  void dispose() {
    _connectionMap.forEach((id, conn) {
      conn.closed = true;
      conn.subscription?.cancel();
      conn.cancelToken?.cancel('disposed');
      conn.reconnectTimer?.cancel();
    });
    _connectionMap.clear();
    _listenMap.clear();
  }

  String _init(String url, bool withCredentials) {
    final id = (_clientId++).toString();
    _connectionMap[id] = _EventSourceConnection();

    final callback = _createCallback();
    final uri = _resolveUri(url);

    _connect(id, uri, withCredentials, callback);
    return id;
  }

  EventSourceEventCallback _createCallback() {
    return (String id, Event event) {
      if (moduleManager?.disposed == true) return;
      moduleManager!.emitModuleEvent(name, event: event, data: id);
    };
  }

  Uri _resolveUri(String input) {
    final Uri parsedUri = Uri.parse(input);
    if (moduleManager != null) {
      Uri base = Uri.parse(moduleManager!.controller.url);
      UriParser uriParser = moduleManager!.controller.uriParser!;
      return uriParser.resolve(base, parsedUri);
    }
    return parsedUri;
  }

  void _connect(String id, Uri uri, bool withCredentials, EventSourceEventCallback callback) {
    final conn = _connectionMap[id];
    if (conn == null || conn.closed) return;

    final useDio = WebFControllerManager.instance.useDioForNetwork;
    if (useDio) {
      _connectWithDio(id, uri, withCredentials, callback);
    } else {
      _connectWithHttpClient(id, uri, withCredentials, callback);
    }
  }

  /// Create a Dio configured for SSE streaming.
  ///
  /// SSE requires a dedicated Dio instance with [IOHttpClientAdapter] because:
  /// - The shared pool may use [CupertinoAdapter] (macOS/iOS) whose underlying
  ///   NSURLSession URLCache can buffer responses before delivering them, which
  ///   blocks indefinitely on infinite SSE streams.
  /// - SSE streams must not be cached or subject to receive timeouts.
  /// - Interceptors from the shared pool (cache, cookies, logging) are not needed
  ///   for SSE and can interfere with streaming delivery.
  Dio _createStreamingDio() {
    final dio = Dio(BaseOptions(
      connectTimeout: const Duration(seconds: 30),
      receiveTimeout: Duration.zero,
      sendTimeout: const Duration(seconds: 60),
      responseType: ResponseType.stream,
    ));
    dio.httpClientAdapter = IOHttpClientAdapter();
    return dio;
  }

  Future<void> _connectWithDio(String id, Uri uri, bool withCredentials, EventSourceEventCallback callback) async {
    final conn = _connectionMap[id];
    if (conn == null || conn.closed) return;

    try {
      final dio = _createStreamingDio();
      final cancelToken = CancelToken();
      conn.cancelToken = cancelToken;

      final headers = <String, dynamic>{
        'Accept': 'text/event-stream',
        'Cache-Control': 'no-cache',
      };
      if (conn.lastEventId != null) {
        headers['Last-Event-ID'] = conn.lastEventId!;
      }

      final response = await dio.getUri<ResponseBody>(
        uri,
        options: Options(
          headers: headers,
          responseType: ResponseType.stream,
          receiveTimeout: Duration.zero,
        ),
        cancelToken: cancelToken,
      );

      if (conn.closed) return;

      // Dispatch 'open' event
      if (_hasListener(id, EVENT_OPEN)) {
        callback(id, Event(EVENT_OPEN));
      }

      // Listen to the stream, parse SSE format.
      // Cast Stream<Uint8List> to Stream<List<int>> for utf8.decoder compatibility.
      final stream = response.data!.stream.cast<List<int>>();
      conn.subscription = stream
          .transform(utf8.decoder)
          .transform(const LineSplitter())
          .listen(
        (line) {
          _processLine(id, line, callback);
        },
        onError: (e) {
          _handleError(id, uri, withCredentials, callback);
        },
        onDone: () {
          _handleDone(id, uri, withCredentials, callback);
        },
      );
    } catch (e, stack) {
      networkLogger.warning('EventSource connect error for $uri', e, stack);
      _handleError(id, uri, withCredentials, callback);
    }
  }

  Future<void> _connectWithHttpClient(
      String id, Uri uri, bool withCredentials, EventSourceEventCallback callback) async {
    final conn = _connectionMap[id];
    if (conn == null || conn.closed) return;

    try {
      final httpClient = createWebFHttpClient();
      final request = await httpClient.getUrl(uri);
      request.headers.set('Accept', 'text/event-stream');
      request.headers.set('Cache-Control', 'no-cache');
      if (conn.lastEventId != null) {
        request.headers.set('Last-Event-ID', conn.lastEventId!);
      }
      conn.httpRequest = request;

      final response = await request.close();

      if (conn.closed) return;

      // Dispatch 'open' event
      if (_hasListener(id, EVENT_OPEN)) {
        callback(id, Event(EVENT_OPEN));
      }

      // HttpClientResponse implements Stream<List<int>>
      conn.subscription = response
          .transform(utf8.decoder)
          .transform(const LineSplitter())
          .listen(
        (line) {
          _processLine(id, line, callback);
        },
        onError: (e) {
          _handleError(id, uri, withCredentials, callback);
        },
        onDone: () {
          _handleDone(id, uri, withCredentials, callback);
        },
      );
    } catch (e, stack) {
      networkLogger.warning('EventSource connect error for $uri', e, stack);
      _handleError(id, uri, withCredentials, callback);
    }
  }

  void _processLine(String id, String line, EventSourceEventCallback callback) {
    final conn = _connectionMap[id];
    if (conn == null || conn.closed) return;

    if (line.isEmpty) {
      // Empty line = dispatch event
      _dispatchEvent(id, conn, callback);
      return;
    }
    if (line.startsWith(':')) return; // Comment, ignore

    String field;
    String value;
    final colonIndex = line.indexOf(':');
    if (colonIndex == -1) {
      field = line;
      value = '';
    } else {
      field = line.substring(0, colonIndex);
      value = line.substring(colonIndex + 1);
      if (value.startsWith(' ')) value = value.substring(1); // Strip leading space per spec
    }

    switch (field) {
      case 'event':
        conn.currentEventType = value;
      case 'data':
        if (conn.dataBuffer.isNotEmpty) conn.dataBuffer.write('\n');
        conn.dataBuffer.write(value);
      case 'id':
        if (!value.contains('\u0000')) conn.lastEventId = value;
      case 'retry':
        final retry = int.tryParse(value);
        if (retry != null) conn.retryMs = retry;
    }
  }

  void _dispatchEvent(String id, _EventSourceConnection conn, EventSourceEventCallback callback) {
    if (conn.dataBuffer.isEmpty) {
      // No data accumulated, reset event type and skip
      conn.currentEventType = 'message';
      return;
    }

    final eventType = conn.currentEventType;
    final data = conn.dataBuffer.toString();
    conn.dataBuffer.clear();
    conn.currentEventType = 'message';

    if (_hasListener(id, eventType)) {
      final event = MessageEvent(data, lastEventId: conn.lastEventId ?? '');
      // Don't override event.type for named events — the bridge uses the type
      // string to choose the JS constructor, and non-standard types like 'update'
      // would produce a plain Event without .data. Instead, encode the named
      // event type in the data parameter so the JS polyfill can re-dispatch.
      final moduleData = eventType == 'message' ? id : '$id\n$eventType';
      callback(moduleData, event);
    }
  }

  void _handleError(String id, Uri uri, bool withCredentials, EventSourceEventCallback callback) {
    final conn = _connectionMap[id];
    if (conn == null || conn.closed) return;

    if (_hasListener(id, EVENT_ERROR)) {
      callback(id, Event(EVENT_ERROR));
    }

    // Schedule reconnection
    _scheduleReconnect(id, uri, withCredentials, callback);
  }

  void _handleDone(String id, Uri uri, bool withCredentials, EventSourceEventCallback callback) {
    final conn = _connectionMap[id];
    if (conn == null || conn.closed) return;

    // Connection closed by server, dispatch error event per spec
    if (_hasListener(id, EVENT_ERROR)) {
      callback(id, Event(EVENT_ERROR));
    }

    // Schedule reconnection
    _scheduleReconnect(id, uri, withCredentials, callback);
  }

  void _scheduleReconnect(String id, Uri uri, bool withCredentials, EventSourceEventCallback callback) {
    final conn = _connectionMap[id];
    if (conn == null || conn.closed) return;

    conn.subscription?.cancel();
    conn.subscription = null;
    conn.cancelToken = null;
    conn.httpRequest = null;

    conn.reconnectTimer = Timer(Duration(milliseconds: conn.retryMs), () {
      if (conn.closed) return;
      _connect(id, uri, withCredentials, callback);
    });
  }

  void _close(String id) {
    final conn = _connectionMap[id];
    if (conn == null) return;

    conn.closed = true;
    conn.subscription?.cancel();
    conn.cancelToken?.cancel('closed');
    conn.reconnectTimer?.cancel();
    _connectionMap.remove(id);
    _listenMap.remove(id);
  }

  bool _hasListener(String id, String type) {
    if (!_listenMap.containsKey(id)) return false;
    return _listenMap[id]!.containsKey(type);
  }

  void _addEvent(String id, String type) {
    if (moduleManager?.disposed == true) return;

    if (!_listenMap.containsKey(id)) {
      _listenMap[id] = {};
    }
    _listenMap[id]![type] = true;
  }
}

File: bridge/polyfill/src/event-source.ts
/*
 * Copyright (C) 2024-present The OpenWebF Company. All rights reserved.
 * Licensed under GNU GPL with Enterprise exception.
 */

import {webf} from './webf';

const esClientMap: Record<string, EventSource> = {};
// Separate storage for named SSE event listeners. The bridge cannot create
// a MessageEvent with a non-standard type (like 'update'), so we route named
// events outside of the normal EventTarget.dispatchEvent path.
const namedListenersMap: Record<string, Record<string, Array<EventListener>>> = {};

function initPropertyHandlersForEventTargets(eventTarget: any, builtInEvents: string[]) {
  for (let i = 0; i < builtInEvents.length; i++) {
    const eventName = builtInEvents[i];
    const propertyName = 'on' + eventName;
    Object.defineProperty(eventTarget, propertyName, {
      get: function () {
        return this['_' + propertyName];
      },
      set: function (value) {
        if (value == null) {
          this.removeEventListener(eventName, this['_' + propertyName]);
        } else {
          this.addEventListener(eventName, value);
        }
        this['_' + propertyName] = value;
      }
    });
  }
}

function validateUrl(url: string): string {
  // Resolve relative URLs
  let resolvedUrl: string;
  try {
    resolvedUrl = new URL(url, location.href).toString();
  } catch (e) {
    throw new SyntaxError(`Failed to construct 'EventSource': The URL '${url}' is invalid.`);
  }
  const protocol = resolvedUrl.substring(0, resolvedUrl.indexOf(':'));
  if (protocol !== 'http' && protocol !== 'https') {
    throw new SyntaxError(
      `Failed to construct 'EventSource': The URL's scheme must be either 'http' or 'https'. '${protocol}' is not allowed.`
    );
  }
  return resolvedUrl;
}

const builtInEvents = ['open', 'message', 'error'];

export class EventSource extends EventTarget {
  static CONNECTING = 0;
  static OPEN = 1;
  static CLOSED = 2;

  CONNECTING: number;
  OPEN: number;
  CLOSED: number;

  url: string;
  withCredentials: boolean;
  readyState: number;
  id: string;

  constructor(url: string | URL, eventSourceInitDict?: { withCredentials?: boolean }) {
    // @ts-ignore
    super();
    this.CONNECTING = EventSource.CONNECTING;
    this.OPEN = EventSource.OPEN;
    this.CLOSED = EventSource.CLOSED;

    const urlStr = typeof url === 'object' && url instanceof URL ? url.toString() : url;
    this.url = validateUrl(urlStr);
    this.withCredentials = eventSourceInitDict?.withCredentials ?? false;
    this.readyState = EventSource.CONNECTING;
    this.id = webf.invokeModule('EventSource', 'init', this.url, this.withCredentials);
    esClientMap[this.id] = this;
    initPropertyHandlersForEventTargets(this, builtInEvents);
  }

  addEventListener(type: string, callback: EventListener | EventListenerObject) {
    webf.invokeModule('EventSource', 'addEvent', this.id, type);
    if (builtInEvents.indexOf(type) === -1) {
      // Named event — store in separate map since the bridge cannot create
      // a MessageEvent with a non-standard type.
      if (!namedListenersMap[this.id]) namedListenersMap[this.id] = {};
      if (!namedListenersMap[this.id][type]) namedListenersMap[this.id][type] = [];
      namedListenersMap[this.id][type].push(callback as EventListener);
    } else {
      super.addEventListener(type, callback);
    }
  }

  removeEventListener(type: string, callback: EventListener | EventListenerObject) {
    if (builtInEvents.indexOf(type) === -1) {
      const listeners = namedListenersMap[this.id]?.[type];
      if (listeners) {
        const idx = listeners.indexOf(callback as EventListener);
        if (idx !== -1) listeners.splice(idx, 1);
      }
    } else {
      super.removeEventListener(type, callback);
    }
  }

  close() {
    this.readyState = EventSource.CLOSED;
    webf.invokeModule('EventSource', 'close', this.id);
    delete esClientMap[this.id];
    delete namedListenersMap[this.id];
  }
}

webf.addWebfModuleListener('EventSource', function (event: any, data: string) {
  // For named SSE events, data is encoded as "clientId\nnamedEventType".
  // For standard events (open, error, message), data is just the clientId.
  let clientId = data;
  let namedEventType = '';
  const nlIdx = data.indexOf('\n');
  if (nlIdx !== -1) {
    clientId = data.substring(0, nlIdx);
    namedEventType = data.substring(nlIdx + 1);
  }

  const client = esClientMap[clientId];
  if (client) {
    switch (event.type) {
      case 'open':
        client.readyState = EventSource.OPEN;
        break;
      case 'error':
        if (client.readyState !== EventSource.CLOSED) {
          client.readyState = EventSource.CONNECTING;
        }
        break;
    }

    if (namedEventType) {
      // Named SSE events arrive as MessageEvent with type='message' from the
      // bridge (to preserve .data). The bridge cannot create a MessageEvent
      // with a non-standard type, and event.type is read-only on the JS side.
      // Dispatch directly to stored named listeners.
      const listeners = namedListenersMap[clientId]?.[namedEventType];
      if (listeners) {
        for (let i = 0; i < listeners.length; i++) {
          listeners[i](event);
        }
      }
    } else {
      client.dispatchEvent(event);
    }
  }
});
